[GTK][WPE] Simplify DRM devices handling

carlosgcampos opened the pull request:
#### 4d8618153bdc635a2cc51f0f4876653d530346f9
<pre>
[GTK][WPE] Simplify DRM devices handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=296929">https://bugs.webkit.org/show_bug.cgi?id=296929</a>

Reviewed by Nikolas Zimmermann.

Add DRMDevice helper struct containing a primary and render node and use
it everywhere instead of specific DRM node paths. This way the
DRMDeviceManager doesn&apos;t need to scan the devices again to find the
nodes, it receives a DRMDevice with the nodes already set.

Canonical link: <a href="https://commits.webkit.org/298556@main">https://commits.webkit.org/298556@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ba2a2a4bc03a9c52eae8527f467fa082fe2e370

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114787 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34532 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24996 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120949 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65472 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116676 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35161 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43090 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87242 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42112 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; layout-tests (exception)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117735 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28012 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103073 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67632 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27197 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21195 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64617 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97396 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21309 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124150 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41791 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31220 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96050 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42169 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99263 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95835 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41015 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18858 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37846 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18520 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41672 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47186 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41234 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44549 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42978 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/gbm/DRMDevice.h:
```diff
@@ -0,0 +1,47 @@
+struct DRMDevice {
+    bool isNull() const
+    {
+        return primaryNode.isNull();
```
I assume this implies renderNode being null to?

carlosgcampos replied:
Yes, a DRM device always have a primary node and optionally a render node too, but I don't think it's possible to have a render node without a primary node.

nikolaszimmermann commented on Source/WebKit/UIProcess/gtk/AcceleratedBackingStore.cpp:
```diff
@@ -527,16 +530,18 @@ void AcceleratedBackingStore::BufferEGLImage::release()
-    auto* device = manager.mainGBMDeviceNode(DRMDeviceManager::NodeType::Render);
-    if (!device) {
+    if (!manager.isInitialized()) {
+        DRMDevice device = drmMainDevice();
```
auto?

nikolaszimmermann commented on Source/WebKit/UIProcess/wpe/WebPageProxyWPE.cpp:
```diff
@@ -155,12 +155,8 @@ Vector<RendererBufferFormat> WebPageProxy::preferredBufferFormats() const
         if (!targetDevice)
             targetDevice = mainDevice;
-        if (targetDevice) {
-            if (bufferFormat.usage == RendererBufferFormat::Usage::Scanout || !wpe_drm_device_get_render_node(targetDevice))
```
Great that this is no longer necessary at this point.

nikolaszimmermann reviewed (approved):
Really nice cleanup! r=me

carlosgcampos committed:
[GTK][WPE] Simplify DRM devices handling
https://bugs.webkit.org/show_bug.cgi?id=296929

Reviewed by Nikolas Zimmermann.

Add DRMDevice helper struct containing a primary and render node and use
it everywhere instead of specific DRM node paths. This way the
DRMDeviceManager doesn't need to scan the devices again to find the
nodes, it receives a DRMDevice with the nodes already set.

Canonical link: https://commits.webkit.org/298556@main
